scripts/base_network.py

```diff
@@ -722,14 +723,297 @@ def _set_shapes(n, country_shapes, offshore_shapes):
+    port (int): The index of the port to extract coordinates from. Default is 0.
+
+    Returns:
+    tuple: The coordinates of the specified port as a tuple (x, y).
+    """
+    linestring = shapely.wkt.loads(linestring_wkt)
+    coords = np.asarray(linestring.coords)
+    ind = [0, -1]
+    coords = coords[ind]
+    coords = coords[port]
+    return coords
+
+
+def _find_closest_lines(lines, new_lines, distance_upper_bound=0.1):
+    """
+    Find the closest lines in a given set of lines to a set of new lines.
+
+    Parameters:
+    lines (pandas.DataFrame): DataFrame with column geometry containing the existing lines.
+    new_lines (pandas.DataFrame): DataFrame with column geometry containing the new lines.
+    distance_upper_bound (float, optional): Maximum distance to consider a line as a match. Defaults to 0.1 which corresponds to approximately 15 km.
+
+    Returns:
+    pandas.Series: Series containing with index the new lines and values providing closest existing line.
+    """
+
+    # get coordinates of start and end points of all lines, for new lines we need to check both directions
+    treelines = lines.geometry.apply(reduce_linestring_to_endpoints)
+    querylines = pd.concat(
+        [
+            new_lines.geometry.apply(reduce_linestring_to_endpoints),
+            new_lines.geometry.apply(reduce_linestring_to_endpoints, reversed=True),
+        ]
+    )
+    treelines = np.vstack(treelines)
+    querylines = np.vstack(querylines)
+    tree = spatial.KDTree(treelines)
+    dist, ind = tree.query(querylines, distance_upper_bound=distance_upper_bound)
+    found_b = ind < len(lines)
+    # since the new lines are checked in both directions, we need to find the correct index of the new line
+    found_i = np.arange(len(querylines))[found_b] % len(new_lines)
+    if len(found_i) < len(new_lines):
+        not_found = new_lines.index.difference(new_lines.index[found_i])
+        logger.warning(
+            "Could not find lines close enough to provided lines:\n"
+            + str(not_found.to_list())
+        )
+    # create a DataFrame with the distances, new line and its closest existing line
+    line_map = pd.DataFrame(
+        dict(D=dist[found_b], i=lines.index[ind[found_b] % len(lines)]),
+        index=new_lines.index[found_i].rename("new_lines"),
+    )
+    # only keep the closer line of the new line pair (since lines are checked in both directions)
+    line_map = line_map.sort_values(by="D")[
+        lambda ds: ~ds.index.duplicated(keep="first")
+    ].sort_index()["i"]
+    return line_map
+
+
+def _adjust_decommissing(branch, n, upgraded_lines, line_map):
+    """
+    Adjust the decommissioning year of the existing lines to the built year of
+    the upgraded lines.
+    """
+    n.df(branch).loc[
+        line_map, "build_year"
+    ] = 2015  # dummy build_year to make existing lines decommissioned when upgraded lines are built
```

Comment:
Is there a way to not hard-code this?

Reply:
Don't know any other method. However, I think we should set it if we keep it that way to an earlier year, maybe 1990.